### PR TITLE
Add static transform component

### PR DESCRIPTION
**Pull request by allenh1**
Adds a static transform publisher component.

This is still in progress, but I'm filing this now for visibility.

@tfoote since these are new files, should I license them as Apache-2.0?

### CODE REVIEW

**Comment by allenh1** [comment]
@clalancette do you think it would be worth using this in the `static_transform_publisher_program.cpp`, or leave that as is?

Definitely need to add some tests still.

**Comment by allenh1** [comment]
@clalancette / @tfoote Could one of you please spin up CI for me?

I saw there are already tests for static transform publisher, and just refactored the existing one to use the component. Tests pass locally, at least.

Let me know what you think!

**Comment by tfoote** [comment]
> @tfoote since these are new files, should I license them as Apache-2.0?

For consistency with the code base please license them under BSD-3 clause. Otherwise you end up requiring double due diligence for the package. 

I have't had a chance to read through it yet, but kicking off CI to see the results.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8511/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4395/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6927/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8420/)

**Comment by allenh1** [comment]
> I haven't had a chance to read through it yet, but kicking off CI to see the results.

@tfoote thanks, found an unfortunate mistake in my use of the logger macros.

Fixed now, sorry I failed to notice that previously.

File: tf2_ros/src/static_transform_broadcaster_program.cpp
Comment (clalancette) [inline tf2_ros/src/static_transform_broadcaster_program.cpp]: This `if` shouldn't be necessary; there isn't any way to get here where the node would be null.  So I'd just remove this.

File: tf2_ros/include/tf2_ros/static_transform_component.hpp
Comment (clalancette) [inline tf2_ros/include/tf2_ros/static_transform_component.hpp]: The style in this repository is generally CamelCase for class names, so I'd change that here.  Also, to bikeshed a bit, I don't love the name `static_transform_component`.  I think something more like `StaticTransformPublisher` is better, since being a component is an implementation detail.  (If you change this, also remember to rename the files).

File: tf2_ros/include/tf2_ros/static_transform_component.hpp
Comment (clalancette) [inline tf2_ros/include/tf2_ros/static_transform_component.hpp]: I don't think that these prototypes can actually be called in the component case.  How would we initialize them in that case?  We may have to think about using parameters instead here.

**Review by clalancette — CHANGES_REQUESTED** [review]
I've got some things that I think need to change here.

File: tf2_ros/src/static_transform_broadcaster_program.cpp
Comment (allenh1) [reply tf2_ros/src/static_transform_broadcaster_program.cpp]: True. Consider it gone.

File: tf2_ros/include/tf2_ros/static_transform_component.hpp
Comment (allenh1) [reply tf2_ros/include/tf2_ros/static_transform_component.hpp]: >  How would we initialize them in that case? 

That's a valid point. I added the `send_transform` function as an attempt to cover this case, but I'm not convinced that's valid.

Those cases are somewhat useful for use with an executor, so I'd like to avoid removing them entirely, but I am not sure what the right thing to do here is.

I'll do the other cleanup items while we figure out the best way to deal with this.

**Comment by allenh1** [comment]
From [here](https://github.com/ros2/geometry2/pull/182#discussion_r337650530):

>  We may have to think about using parameters instead here.

My concern with parameters is how that would be done in a system with several of these components. If they all use something like "linear_x" for the name, how does one keep the parameters from getting clobbered by other static transform setups?

I'm really not sure how to support the generic component usage.

**Comment by clalancette** [comment]
> 
> My concern with parameters is how that would be done in a system with several of these components. If they all use something like "linear_x" for the name, how does one keep the parameters from getting clobbered by other static transform setups?

The individual nodes would have to be named differently or namespaced differently, which would solve that problem.  Even though we don't currently enforce it, you really want to do that anyway, since there is currently no other way to uniquely identify a node.

**Comment by allenh1** [comment]
> The individual nodes would have to be named differently or namespaced differently, which would solve that problem

So I should probably take the `name` parameter and append it so it would look like this?

```
/[name]/translation/x
/[name]/translation/y
/[name]/translation/z
/[name]/rotation/x
/[name]/rotation/y
/[name]/rotation/z
/[name]/rotation/w
```

**Comment by clalancette** [comment]
> So I should probably take the `name` parameter and append it so it would look like this?

Honestly, I'd get rid of the `name` parameter altogether.  This is no different than if you try to run multiple of *any* kind of node; you need to uniquely name them one way or another.

**Comment by allenh1** [comment]
> Honestly, I'd get rid of the name parameter altogether

Okay, so, to make sure I understand,  you want the above but without the leading `/name`, which would be remapped by the user.

**Comment by allenh1** [comment]
@clalancette I have not declared the parameters yet, but I'd like you to take a look at things to make sure we're still on the same page if you have a minute.

**Comment by clalancette** [comment]
> @clalancette I have not declared the parameters yet, but I'd like you to take a look at things to make sure we're still on the same page if you have a minute.

So I do like the `declare_parameters`; that's nice.

The ability to dynamically change the parameters on the fly is neat, but I actually don't think it belongs here.  That is, this is the *static* transform broadcaster, so being able to change it on the fly gives me cognitive dissonance.  It also makes it more complicated for an unclear use-case.  So I'd suggest not installing the parameter_callback at all, and just declaring the parameters at constructor time.

**Comment by allenh1** [comment]
> The ability to dynamically change the parameters on the fly is neat, but I actually don't think it belongs here. 

Yeah, that's a good point.

>  So I'd suggest not installing the parameter_callback at all, and just declaring the parameters at constructor time.

As I'm frankly a bit out of the loop here, I'm not quite sure what the "proper" way to do that would be. Is this still `get_parameter` or is there a new mechanism? Maybe [this](https://github.com/ros2/demos/blob/6c4323261c335426af871a4d67f4025752c45417/demo_nodes_cpp/src/parameters/set_and_get_parameters.cpp#L72-L78)?

**Comment by clalancette** [comment]
> > 
> 
> As I'm frankly a bit out of the loop here, I'm not quite sure what the "proper" way to do that would be. Is this still `get_parameter` or is there a new mechanism? Maybe [this](https://github.com/ros2/demos/blob/6c4323261c335426af871a4d67f4025752c45417/demo_nodes_cpp/src/parameters/set_and_get_parameters.cpp#L72-L78)?

Personally I would just take your `declare_parameter` lines (lines 57 - 65), move them to the top of the constructor, and mark them as "read-only" (something like https://github.com/ros2/rclcpp/blob/2716d3e81ed1bfd8b1afe8d07bdeda7d3a4ea61b/rclcpp/test/test_node.cpp#L315).  That way the parameters are still introspectable, but can't be changed.

**Comment by allenh1** [comment]
> Personally I would just take your declare_parameter lines (lines 57 - 65), move them to the top of the constructor, and mark them as "read-only"

Done! Tested and working (at least locally).

@tfoote / @clalancette please review and or spin up CI for me (when you are able)!

File: tf2_ros/CMakeLists.txt
Comment (clalancette) [inline tf2_ros/CMakeLists.txt]: The name of this file (and target) should be changed to `static_transform_publisher`.

File: tf2_ros/include/tf2_ros/static_transform_component.hpp
Comment (clalancette) [inline tf2_ros/include/tf2_ros/static_transform_component.hpp]: I think we should just drop the `name` parameter here; I don't think it is something we can reasonably pass in the component case anyway.

File: tf2_ros/include/tf2_ros/static_transform_component.hpp
Comment (clalancette) [inline tf2_ros/include/tf2_ros/static_transform_component.hpp]: The only include files you need in the header file are:

* `memory`
* `rclcpp/rclcpp.hpp`
* `tf2_ros/tf2_ros.h`

The rest of them can be removed.

File: tf2_ros/include/tf2_ros/static_transform_component.hpp
Comment (clalancette) [inline tf2_ros/include/tf2_ros/static_transform_component.hpp]:
```suggestion
  std::unique_ptr<tf2_ros::StaticTransformBroadcaster> broadcaster_;
```

File: tf2_ros/include/tf2_ros/static_transform_component.hpp
Comment (clalancette) [inline tf2_ros/include/tf2_ros/static_transform_component.hpp]:
```suggestion
#ifndef TF2_ROS__STATIC_TRANSFORM_PUBLISHER_HPP_
```

File: tf2_ros/src/static_transform_broadcaster_program.cpp
Comment (clalancette) [inline tf2_ros/src/static_transform_broadcaster_program.cpp]: Since you are adding usage of a `std::shared_ptr`, add a `#include <memory>` at the top of the file.

File: tf2_ros/src/static_transform_broadcaster_program.cpp
Comment (clalancette) [inline tf2_ros/src/static_transform_broadcaster_program.cpp]: I think this `TODO` should move inside of the `StaticTransformPublisher` constructor, since that is where we would do it now.

File: tf2_ros/src/static_transform_component.cpp
Comment (clalancette) [inline tf2_ros/src/static_transform_component.cpp]: Make sure to include what you use here:

```
#include <chrono>
#include <memory>
```

File: tf2_ros/src/static_transform_component.cpp
Comment (clalancette) [inline tf2_ros/src/static_transform_component.cpp]: I don't think this dance is necessary.  Looking at the [code](https://github.com/ros2/rclcpp/blob/2716d3e81ed1bfd8b1afe8d07bdeda7d3a4ea61b/rclcpp/src/rclcpp/node_interfaces/node_clock.cpp#L33), I think the default clock on a node is already an `RCL_ROS_TIME`, so down below you can just call `this->now()`.

File: tf2_ros/src/static_transform_component.cpp
Comment (clalancette) [inline tf2_ros/src/static_transform_component.cpp]: Instead of `rclcpp::shutdown` and a return here, I think an exception is more in order.  At least that will show up in the component container.

File: tf2_ros/src/static_transform_component.cpp
Comment (clalancette) [inline tf2_ros/src/static_transform_component.cpp]: As mentioned above, just call `this->now()` here.

**Review by clalancette — CHANGES_REQUESTED** [review]
Got a bunch of comments.  They should be mostly easy to fix.

File: tf2_ros/CMakeLists.txt
Comment (allenh1) [inline tf2_ros/CMakeLists.txt]: @clalancette errr seems `static_transform_publisher` is taken... What would you prefer in its place?

Comment (clalancette) [reply tf2_ros/CMakeLists.txt]: Oh, haha.  I didn't even realize.

Looking at this again, what this ends up being is a node wrapping the `StaticTransformBroadcaster` class.  So how about naming the class `StaticTransformBroadcasterNode`, and the file `static_transform_broadcaster_node.cpp`?

**Comment by allenh1** [comment]
@clalancette ready for round 3!

---

Also, should I be worried about a failed build here? It seems pretty consistently red, but I got a green checkmark on my penultimate commit so I'm a bit confused.

**Comment by clalancette** [comment]
@allenh1 I made a few changes here.  In particular:

1.  I changed the constructor to take a const ref `rclcpp::NodeOptions`
1.  I simplified the constructor, since we can both `declare_parameter` and get the value at the same time.
1.  I fixed up the includes slightly more.

I'm now happy with this, so I'll give it an approval, but I'd appreciate if you could take a look at what I changed and tell me whether you agree with it.

**Review by clalancette — APPROVED** [review]

**Comment by allenh1** [comment]
@clalancette yep -- those changes look good to me! Thanks.

Would you please spin up CI?

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8547/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4418/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6967/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8479/)

**Comment by allenh1** [comment]
@clalancette Ah, yes, my good friend the windows linker spoiled our fun. Constructor for the node is now public in the DLL.

**Comment by clalancette** [comment]
@allenh1 Of course :).  I just re-kicked the Windows CI job, let's see if it is happier now.

**Comment by allenh1** [comment]
> let's see if it is happier now.

@clalancette seems happier, but still not happy. It's complaining about inconsistent DLL linkage. Any idea why that would be?

**Comment by clalancette** [comment]
I just tried something different here, which is to make the *class* public, not just the constructor.  Let's see what happens.

(I also rebased it on the latest)

**Comment by allenh1** [comment]
@clalancette aghhh.... Red build again.

It seems to be complaining because `rclcpp::Node` is not public in the dll?

**Comment by clalancette** [comment]
@allenh1 I give up.  I'd need to spend some time on Windows to figure it out, but I won't have time for that until after ROSCon.  If you have some time and a Windows machine you can try on, that would be great.

**Comment by clalancette** [comment]
Thanks to @sloretz, I think I've figured this out for real.  I'll kick off another CI here and check again tomorrow.

**Comment by allenh1** [comment]
@clalancette / @sloretz thanks for the help! I definitely learned something here, haha. Strange stuff!

**Comment by allenh1** [comment]
@clalancette CI results from above, posted again since they are hidden:

CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8547/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4418/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6967/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8479/)

**Comment by allenh1** [comment]
@clalancette (sorry to tag you a third time -- random side thought) I'm wondering if it might be useful to create a generator for that file? I hesitate to file an issue somewhere because, frankly, I'm really not sure if that should be an `rclcpp` thing, `rcl` thing, `ament` thing (my favorite of the list), or something else entirely.

To me, that file looks super repetitive and just begs to be generated.

**Comment by clalancette** [comment]
> To me, that file looks super repetitive and just begs to be generated.

It turns out that there is a generator for it in CMake, but we've been somewhat reluctant to use it in the ROS 2 core.  Mostly because if you use it, then you have to build with CMake.  This obviously isn't a problem for us, but there are some groups that may want to build using other build systems.

I'm not really going to weigh in on the pros or cons of that approach in this PR.  I think what we have now will work and is good, so with green CI I'm happy with it.

Because we are in Eloquent freeze right now, though, we'll have to wait to merge this until after we create an eloquent branch.

**Comment by allenh1** [comment]
> Because we are in Eloquent freeze right now, though, we'll have to wait to merge this until after we create an eloquent branch.

Good thing I didn't smash that squash and merge button then, haha.

Thanks for all the iterating! I'll watch for the eloquent branch and ping you again once I see it.

**Comment by clalancette** [comment]
@allenh1 Why the changes in 15a8df9?  I'm pretty sure the node was already being installed into lib, and it is not a library so doesn't need to be exported.

**Comment by allenh1** [comment]
> @allenh1 Why the changes in 15a8df9? I'm pretty sure the node was already being installed into lib, and it is not a library so doesn't need to be exported.

@clalancette it was being installed into lib, but I wanted to use the component in a different package, which could not find the library. On windows, I could find the `.dll` file, but I could not find the `.lib` file.

Is this an invalid use case for components? If so I can just drop that commit.

File: tf2_ros/src/static_transform_broadcaster_node.cpp
Comment (allenh1) [inline tf2_ros/src/static_transform_broadcaster_node.cpp]: I feel like I should also adjust this to be `1.0`, since the default quaternion is not valid here.

**Comment by clalancette** [comment]
> @clalancette it was being installed into lib, but I wanted to use the component in a different package, which could not find the library. On windows, I could find the `.dll` file, but I could not find the `.lib` file.

No, sorry, I was just getting confused with the names.  Sorry about that.  That change is fine.

File: tf2_ros/src/static_transform_broadcaster_node.cpp
Comment (clalancette) [reply tf2_ros/src/static_transform_broadcaster_node.cpp]: Yeah, that's a great idea.

**Comment by allenh1** [comment]
@clalancette While we're waiting on the branch, I'd like to bring up an observation from using this.

I get a warning in the console when I use multiple instances of this component due to the name of the node being duplicated.

Should I add a `static` counter to the class that will increment when the constructor is called, then append the counter to the name of the node? I think that makes the node names look rather silly, but would solve this issue.

Does this have to do with the `TODO` about anonymizing the name of the node?

**Comment by allenh1** [comment]
@clalancette / @tfoote rebased on top of `ros2` -- should CI be run on this one more time?

**Comment by allenh1** [comment]
@clalancette since the eloquent branch is now available and this is rebased on the tip of ros2, is there anything else I should wait for before hitting that squash and merge?

Maybe one more round of CI since I'm paranoid (which you'll need to start for me, of couse ;) ).

**Comment by clalancette** [comment]
@allenh1 Now that we are past freeze, we can reconsider this one.

First, do you mind squashing all of this development into a single commit?  It's pretty messy to look at (a lot of that is my fault, but we may as well fix it now).

Second, I think your point about the name of the node is a really good one.  Yes, that is part of what anonymizing the name would buy us.  We should think about how to fix it before we merge this, as that will bite people pretty easily.

The problem with adding a static counter to the class is that it only helps *for this process*.  If you launch two processes, and they each have a static_transform_publisher in them, they'll still overlap.  One way to go about this (without true anonymous name support) is to generate a UUID for the node and suffix that onto the name.  See https://github.com/ros2/rclcpp/blob/e494b3efada0258845e41ca866c5baadc8c9a83c/rclcpp_action/src/client.cpp#L370 , where we do something similar for actions.  I'll propose we do something similar for now, and we can always update it later to do something smarter if we get true anonymous node support.

**Comment by allenh1** [comment]
>  It's pretty messy to look at (a lot of that is my fault, but we may as well fix it now).

It was quite a bit more iteration than I was also anticipating, haha. Squashed. Now we can pretend we just got it right a lot earlier on ;).

> The problem with adding a static counter to the class is that it only helps for this process

That's a valid point.

> One way to go about this (without true anonymous name support) is to generate a UUID for the node and suffix that onto the name.

I also think this is the best approach (at least for the time being). I'll get to that soon.

File: tf2_ros/src/static_transform_broadcaster_node.cpp
Comment (allenh1) [inline tf2_ros/src/static_transform_broadcaster_node.cpp]: @clalancette I'm not super proud of this, but it works... Do you have any other recommendations?

Comment (clalancette) [reply tf2_ros/src/static_transform_broadcaster_node.cpp]: @allenh1 It's OK, but I feel like it is overly complicated and not performant enough (the latter doesn't matter *too* much here, but still).  I played around with this a bit and came up with: https://gist.github.com/clalancette/fedda4179d8956a13bb4e782f8c09dbc ; what do you think about that?

Comment (allenh1) [reply tf2_ros/src/static_transform_broadcaster_node.cpp]: >  I feel like it is overly complicated and not performant enough

Yep. 100% agreed.

> what do you think about that?

I like your implementation better. Swapped it in.

**Comment by allenh1** [comment]
@clalancette could you spin up CI for me please?

**Comment by clalancette** [comment]
> @clalancette could you spin up CI for me please?

Sure!  Here you go:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8808/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4624/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7222/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8725/)

**Comment by allenh1** [comment]
> Sure! Here you go

@clalancette Hm... Seems something is not quite right. I ran `colcon test --packages-select tf2_ros` locally, then `colcon test-result` and got the following on my machine:

```
G:\dashing_ws>colcon test-result
c:\python37\lib\site-packages\colcon_cmake\test_result\ctest.py:62: DeprecationWarning: This method will be removed in future versions.  Use 'list(elem)' or iteration over elem instead.
  children = root.getchildren()
Summary: 42 tests, 0 errors, 0 failures, 0 skipped
```

Did I need to run a different test?

**Comment by clalancette** [comment]
> @clalancette Hm... Seems something is not quite right. I ran `colcon test --packages-select tf2_ros` locally, then `colcon test-result` and got the following on my machine:

I'm not sure why you are getting a different result than CI.  However, I will point out that the tests that failed in CI here also failed on the nightlies: https://ci.ros2.org/view/nightly/job/nightly_linux_debug/1398/#showFailuresLink .  So it's not the fault of this PR.

At this point, I'm pretty good with this.  I'll let it sit until Monday (I'm pretty busy today anyway), but if I don't get other opinions by then I'll merge it.

**Comment by allenh1** [comment]
>  I'll let it sit until Monday (I'm pretty busy today anyway), but if I don't get other opinions by then I'll merge it.

🎉 Thanks!

> I'm not sure why you are getting a different result than CI. 

Me either... Just a classic "works on my machine" I suppose.

**Pull request merged**
